File: solutions/clojure/game-of-life/1/src/game_of_life.clj
(ns game-of-life)

(defn- neighbour-cells [[row col]]
  (for [a [-1 0 1] b [-1 0 1]
        :when (not= [0 0] [a b])]
    [(+ row a) (+ col b)]))


(defn- all-positions [matrix]
  (for [row (range 0 (count matrix))
        col (range 0 (count (first matrix)))]
    [row col]))


(defn- alive? [gen cell]
  (= 1 (get-in gen cell)))


(defn- ->as-alive-set [cells]
  (->> cells
       all-positions
       (filter #(alive? cells %))
       set))

(defn- cells-range [cells]
  [(count cells) (count (first cells))])

(defn- in-range? [[rows cols] [row col]]
  (and (<= 0 row (dec rows)) (<= 0 col (dec cols))))


(defn- ->cells-matrix [[rows cols] alive-cell?]
  (mapv
    #(mapv
      (fn [y] (if (alive-cell? [% y]) 1 0))
      (range cols))
    (range rows)))


(defn- alive-next-gen? [n was-alive?]
  (or (= n 3) (and (= n 2) was-alive?)))


(defn tick
  "Returns the next generation of the cells."
  [cells]
  (let [rng (cells-range cells)
        alive-cells (->as-alive-set cells)]
    (->> (for [[cell n] (frequencies (mapcat neighbour-cells alive-cells))
               :when (alive-next-gen? n (alive-cells cell))]
           cell)
         (filter (partial in-range? rng))
         set
         (->cells-matrix rng))))

File: solutions/python/game-of-life/1/game_of_life.py
"""Calculate the next generation in Conway's Game of Life."""

from collections import Counter

ALIVE = 1
NEIGHBOR_DELTAS = (
    (-1, -1), (-1, 0), (-1, 1),
    (0, -1), (0, 1),
    (1, -1), (1, 0), (1, 1),
)


def tick(matrix):
    """Return the next generation of the matrix."""
    alive_cells = _alive_cells(matrix)
    neighbor_counts = _neighbor_counts(alive_cells, matrix)

    return [
        [
            int(_alive_next_generation((row, column), alive_cells, neighbor_counts))
            for column in range(_width(matrix))
        ]
        for row in range(_height(matrix))
    ]


def _alive_cells(matrix):
    return {
        (row, column)
        for row, cells in enumerate(matrix)
        for column, value in enumerate(cells)
        if value == ALIVE
    }


def _neighbor_counts(alive_cells, matrix):
    return Counter(
        neighbor
        for cell in alive_cells
        for neighbor in _neighbors(cell, matrix)
    )


def _alive_next_generation(cell, alive_cells, neighbor_counts):
    live_neighbors = neighbor_counts[cell]

    return live_neighbors == 3 or (cell in alive_cells and live_neighbors == 2)


def _neighbors(cell, matrix):
    row, column = cell

    return (
        (row + row_delta, column + column_delta)
        for row_delta, column_delta in NEIGHBOR_DELTAS
        if _in_bounds(matrix, row + row_delta, column + column_delta)
    )


def _in_bounds(matrix, row, column):
    return 0 <= row < _height(matrix) and 0 <= column < _width(matrix)


def _height(matrix):
    return len(matrix)


def _width(matrix):
    return len(matrix[0]) if matrix else 0
